internal/controller/manifests_downloader.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	operatorv1 "sigs.k8s.io/cluster-api-operator/api/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	configMapVersionLabel = "provider.cluster.x-k8s.io/version"
+	configMapTypeLabel    = "provider.cluster.x-k8s.io/type"
+	configMapNameLabel    = "provider.cluster.x-k8s.io/name"
+	operatorManagedLabel  = "managed-by.operator.cluster.x-k8s.io"
+
+	metadataConfigMapKey   = "metadata"
+	componentsConfigMapKey = "components"
+)
+
+// downloadManifests downloads CAPI manifests from a url.
+func (p *phaseReconciler) downloadManifests(ctx context.Context) (reconcile.Result, error) {
+	// Return immediately if a custom config map is used instead of a url.
+	if p.provider.GetSpec().FetchConfig != nil && p.provider.GetSpec().FetchConfig.Selector != nil {
+		return reconcile.Result{}, nil
+	}
+
+	// Check if manifests are already downloaded and stored in a configmap
+	labelSelector := metav1.LabelSelector{
+		MatchLabels: p.prepareConfigMapLabels(),
+	}
+
+	exists, err := p.checkConfigMapExists(ctx, labelSelector)
+	if err != nil {
+		return reconcile.Result{}, wrapPhaseError(err, "failed to check that config map with manifests exists", operatorv1.PreflightCheckCondition)
+	}
+
+	if exists {
+		return reconcile.Result{}, nil
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+
+	log.Info("Downloading provider manifests")
```

Comment:
we could add a verbosity to the logger:
```suggestion
	log.V(5).Info("Downloading provider manifests")
```